google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/BigtableBatchingCallSettings.java

```diff
@@ -145,9 +193,106 @@ public Builder setRetrySettings(@Nonnull RetrySettings retrySettings) {
+      this.targetRpcLatencyMs = isLatencyBasedThrottlingEnabled ? targetRpcLatency : null;
+      return this;
+    }
+
+    /** Gets target rpc latency if latency based throttling is enabled. Otherwise returns null. */
+    @Nullable
+    public Long getTargetRpcLatencyMs() {
+      return isLatencyBasedThrottlingEnabled ? targetRpcLatencyMs : null;
+    }
+
+    /** Gets if latency based throttling is enabled. */
+    public boolean isLatencyBasedThrottlingEnabled() {
+      return this.isLatencyBasedThrottlingEnabled;
+    }
+
+    /**
+     * Gets {@link DynamicFlowControlSettings}.
+     *
+     * <p>By default, DynamicFlowControlSettings uses settings in {@link
+     * BatchingSettings#getFlowControlSettings()}. If maxOutstandingElementCount is not set in
+     * {@link BatchingSettings#getFlowControlSettings()}, maxOutstandingElementCount =
+     * Math.min(20000, 2000 * number of runtime processors). If maxOutstandingRequestBytes is not
+     * set in {@link BatchingSettings#getFlowControlSettings()}, maxOutstandingRequestBytes = 100MB.
+     *
+     * <p>If latency based throttling is enabled, initialOutstandingElementCount = Math.max(batch
+     * element count, maxOutstandingElementCount / 4); minOutstandingElementCount = Math.max(batch
+     * element count, maxOutstandingElementCount / 100). If latency based throttling is disabled,
+     * initialOutstandingElementCount, maxOutstandingElementCount and minOutstandingElementCount
+     * will be the same.
+     *
+     * <p>Latency based throttling only adjusts outstanding element count.
+     * initialOutstandingRequestBytes, minOutstandingRequestBytes and maxOutstandingRequestBytes
+     * will always be the same.
+     *
+     * <p>{@link LimitExceededBehavior} will always be {@link LimitExceededBehavior#Block}.
+     */
+    public DynamicFlowControlSettings getDynamicFlowControlSettings() {
+      return this.dynamicFlowControlSettings;
+    }
+
     /** Builds the {@link BigtableBatchingCallSettings} object with provided configuration. */
     @Override
     public BigtableBatchingCallSettings build() {
+      Preconditions.checkState(batchingSettings != null, "batchingSettings must be set");
+      FlowControlSettings defaultSettings = batchingSettings.getFlowControlSettings();
+      if (isLatencyBasedThrottlingEnabled()) {
+        Long maxThrottlingElementCount = defaultSettings.getMaxOutstandingElementCount();
+        Long maxThrottlingRequestByteCount = defaultSettings.getMaxOutstandingRequestBytes();
+        if (maxThrottlingElementCount == null) {
+          long maxBulkMutateElementPerBatch = 100L;
+          long defaultChannelPoolSize = 2 * Runtime.getRuntime().availableProcessors();
+          maxThrottlingElementCount =
+              Math.min(20_000L, 10L * maxBulkMutateElementPerBatch * defaultChannelPoolSize);
+        }
+        if (maxThrottlingRequestByteCount == null) {
+          maxThrottlingRequestByteCount = 100L * 1024 * 1024;
+        }
```

Comment:
It seems like you are duplicating the defaults from EnhancedBigtableStubSettings. This will be very hard to keep in sync. If these settings are required and the user hasnt set them, then I think we should raise an error.